pkg/config/validator.go

```diff
@@ -0,0 +1,151 @@
+package config
+
+import (
+	"fmt"
+	"github.com/xeipuuv/gojsonschema"
+	"sigs.k8s.io/yaml"
+	"strings"
+)
+
+const (
+	defaultVersion = "1.0"
+	schemaFilePath = "data/config_schema_v%s.json"
+)
+
+func getSchema(version string) (gojsonschema.JSONLoader, error) {
+	if version == "" {
+		version = defaultVersion
+	}
+	schemaData, err := Asset(fmt.Sprintf(schemaFilePath, version))
+	if err != nil {
+		return nil, err
+	}
+	return gojsonschema.NewStringLoader(string(schemaData)), nil
+}
+
+func ValidateConfig(config *Config, version string) error {
+	schemaLoader, err := getSchema(version)
+	if err != nil {
+		return err
+	}
+	dataLoader := gojsonschema.NewGoLoader(config)
+	return ValidateSchema(schemaLoader, dataLoader)
+}
+
+func Validate(yamlConfig string, version string) error {
+	j := []byte(yamlConfig)
+	config, err := yaml.YAMLToJSON(j)
+	if err != nil {
+		return err
+	}
+
+	schemaLoader, err := getSchema(version)
+	if err != nil {
+		return err
+	}
+	dataLoader := gojsonschema.NewStringLoader(string(config))
+	return ValidateSchema(schemaLoader, dataLoader)
+}
+
+func ValidateSchema(schemaLoader, dataLoader gojsonschema.JSONLoader) error {
+	result, err := gojsonschema.Validate(schemaLoader, dataLoader)
+	if err != nil {
+		return err
+	}
+
+	if !result.Valid() {
+		return toError(result)
+	}
+	return nil
+}
+
+func toError(result *gojsonschema.Result) error {
+	err := getMostSpecificError(result.Errors())
+	return err
+}
+
+const (
+	jsonschemaOneOf = "number_one_of"
+	jsonschemaAnyOf = "number_any_of"
+)
+
+func getDescription(err validationError) string {
+	switch err.parent.Type() {
```

Comment:
Some documentation would be nice here. Does this collapse a list of errors into a single error?

Reply:
While validating the schema there can be a lot of errors and it was confusing on what to return. One approach was to return a list of errors.

 I was trying to find a standard way and found this in docker project: 
https://github.com/farfeduc/docker-ce/blob/74c07ebc84244a5df3c25786e477500b446e9704/components/cli/cli/compose/schema/schema.go#L79

This tries to find the most specific error based on the number of times it was repeated.
Also, it tries to find the parent error.